File: go/52.go
/*

https://leetcode.com/problems/n-queens-ii/

The n-queens puzzle is the problem of placing n queens on an n×n chessboard such that no two queens attack each other.



Given an integer n, return the number of distinct solutions to the n-queens puzzle.

Example:

Input: 4
Output: 2
Explanation: There are two distinct solutions to the 4-queens puzzle as shown below.
[
 [".Q..",  // Solution 1
  "...Q",
  "Q...",
  "..Q."],

 ["..Q.",  // Solution 2
  "Q...",
  "...Q",
  ".Q.."]
]

*/

package main


import "fmt"



func validate(board [][]string, row int, col int) bool {
    for i := 0; i < len(board); i++ {
        for j := 0; j < col; j++ {
            if board[i][j] == "Q" {
                if row + j == col + i || row + col == i + j || row == i {
                    return false
                }
            }
        }
    }
    return true
}


func dfs(board [][]string, col int, total *int) {
    if col == len(board) {
        *total = *total + 1
        return
    }
    for row := 0; row < len(board); row++ {
        if validate(board, row, col)  {
            board[row][col] = "Q"
            dfs(board, col + 1, total)
            board[row][col] = "."
        }
    }
}


func totalNQueens(n int) int {
    board := make([][]string, n)
    for i := 0; i < n; i++ {
        board[i] = make([]string, n)
        for j := 0; j < n; j++ {
            board[i][j] = "."
        }
    }

    var total int = 0
    dfs(board, 0, &total)
    return total
}

// 这个虽然跑得慢 但是容易理解 理解才能够应用
// 其他的版本稍微复杂有点 理解起来有点麻烦

func main() {
    for i := 4; i < 10; i++ {
        fmt.Println(i, totalNQueens(i))
    }
}

File: python/52.py
"""
https://leetcode.com/problems/n-queens-ii/

The n-queens puzzle is the problem of placing n queens on an n×n chessboard such that no two queens attack each other.



Given an integer n, return the number of distinct solutions to the n-queens puzzle.

Example:

Input: 4
Output: 2
Explanation: There are two distinct solutions to the 4-queens puzzle as shown below.
[
 [".Q..",  // Solution 1
  "...Q",
  "Q...",
  "..Q."],

 ["..Q.",  // Solution 2
  "Q...",
  "...Q",
  ".Q.."]
]

"""


class Solution:
    def totalNQueens(self, n: int) -> int:
        res = 0

        def f(board, row):
            if row == n:
                nonlocal res
                res += 1
                return
            for col in range(n):
                if check(board, row, col):
                    board[row][col] = 'Q'
                    f(board, row + 1)
                    board[row][col] = '.'

        # 判重的过程需要好好思考啊
        def check(board, row, col):
            # check column
            for i in range(row):
                if board[i][col] == 'Q':
                    return False
            # 因为是按照row从小到大来的, 所以row不会超，判断的范围在(0, row)即可
            # 但是col的返回一开始就是从(1, n), 所以每次判断的范文是(0, n)
            # 就是这个原因, 就是这么简单
            # check 45 degree
            i, j = row - 1, col - 1
            while i >= 0 and j >= 0:
                if board[i][j] == 'Q':
                    return False
                i, j = i - 1, j - 1

            # check 135 degree
            i, j = row - 1, col + 1
            while i >= 0 and j < n:
                if board[i][j] == 'Q':
                    return False
                i, j = i - 1, j + 1

            return True

        board = [['.'] * n for _ in range(n)]
        f(board, 0)
        return res


if __name__ == '__main__':
    for i in range(4, 11):
        print(i, Solution().totalNQueens(i))
